File: shared/src/ws-events.ts
// The app WebSocket contract (the browser client <-> backend socket at /api). Distinct from the
// runner protocol (see runner-protocol.ts). The backend emits ServerEvent frames (some to one
// socket, most fanned out to a channel or broadcast to all); the client sends ClientFrame frames.

import type { AgentStatus, Deliverable, Participant, WireMessage } from "./domain.js";
import type { SlackChannelLink } from "./slack.js";

// ---- Server -> client ----

// Sent to a socket right after it authenticates.
export interface ConnectedEvent {
  type: "connected";
  participantId: string;
}

// Sent to a socket when its last inbound frame could not be processed.
export interface ErrorEvent {
  type: "error";
  error: string;
}

// The reply to a client `ping`. This is the client's ONLY way to prove the socket is still alive:
// the browser WebSocket API answers protocol-level pings itself and never surfaces them to JS, so
// liveness has to ride at the application level. Without it a half-open socket (laptop sleep, wifi
// switch, an idle NAT dropping the flow) sits at readyState OPEN forever, fires no `close`, and
// the tab goes quiet until the user reloads.
export interface PongEvent {
  type: "pong";
}

// A new (or thread-reply) message in a channel the recipient belongs to.
export interface MessageEvent {
  type: "message";
  message: WireMessage;
}

// An agent's live status changed (working/idle/sleeping/waking).
export interface AgentStatusChangedEvent {
  type: "agent_status_changed";
  agentId: string;
  status: AgentStatus;
}

// A self-hosted device's control connection came up or went down. Fanned out to the sockets of
// the device's OWNER account (a device is account-scoped, not workspace-scoped), so an open
// Environments page flips the online dot without a refetch. The agents running on that device
// emit their own agent_status_changed (offline/idle) separately.
export interface DeviceStatusChangedEvent {
  type: "device_status_changed";
  deviceId: string;
  online: boolean;
}

// A channel's membership changed (added/removed member); clients refetch members.
export interface MembersChangedEvent {
  type: "members_changed";
  channelId: string;
}

// A channel was deleted.
export interface ChannelDeletedEvent {
  type: "channel_deleted";
  channelId: string;
}

// A channel was created with the recipient already a member (fanned out to those members).
// Coarse by design, like members_changed: clients refetch their channel list rather than trying
// to splice a row in with the right unread/ordering state. DMs are NOT announced this way — an
// empty DM isn't worth a sidebar row, and the first message in one is what makes it appear (the
// client refetches when a message arrives for a channel it doesn't know).
export interface ChannelCreatedEvent {
  type: "channel_created";
  channelId: string;
}

// A participant's editable fields changed. Two sources: a human saving the profile form, and an
// agent calling set_status (which is why this can fire many times a minute during a busy turn —
// clients replace the roster entry, they don't refetch). Carries the public participant.
export interface ParticipantUpdatedEvent {
  type: "participant_updated";
  participant: Participant;
}

// A participant joined the workspace — an agent someone (or some agent) just created, a human who
// accepted an invite, a Slack shadow user. Broadcast workspace-wide so every open client adds them
// to the roster: the Team page, member pickers, @-autocomplete. The counterpart of
// participant_deleted; without it a brand-new agent is invisible (and un-mentionable) to everyone
// who didn't create it until they refresh. Carries the PUBLIC row (no runner_token).
export interface ParticipantCreatedEvent {
  type: "participant_created";
  participant: Participant;
}

// A participant (agent) was permanently deleted.
export interface ParticipantDeletedEvent {
  type: "participant_deleted";
  participantId: string;
}

// Where an agent's current turn came from: the channel (and thread/message) whose dispatch the
// runner consumed. This is what lets the client show work WHERE IT WAS REQUESTED — the trigger
// message's chip, the DM strip, the sidebar working-dot — instead of in every channel the agent
// happens to be a member of. Absent for turns with no dispatch context (e.g. compaction).
export interface TurnContext {
  channelId?: string;
  threadRootId?: string | null;
  messageId?: string; // the message whose dispatch triggered this turn
}

// A turn began: which agent, which turn, and where it was triggered from.
export interface AgentTurnEvent {
  type: "agent_turn";
  agentId: string;
  turnId: string;
  context: TurnContext | null;
}

// One raw SDK stream message from an agent's turn, for the live Activity transcript. Carries the
// turn's context on every frame so a client that loads mid-turn still learns the turn's home.
export interface AgentEventEvent {
  type: "agent_event";
  agentId: string;
  turnId: string;
  event: unknown;
  context?: TurnContext | null;
}

// A dispatch landed in the agent's inbox behind a turn already in progress — no turn_id yet
// (that only exists once the runner actually starts or splices it in). Lets the triggering
// message show a "queued — waiting for @agent" chip immediately instead of nothing.
export interface AgentQueuedEvent {
  type: "agent_queued";
  agentId: string;
  context: TurnContext;
}

// An agent's context-window occupancy after a turn (drives the profile usage meter).
export interface AgentContextEvent {
  type: "agent_context";
  agentId: string;
  tokens: number;
  maxTokens: number;
}

// An agent's long-term memory (MEMORY.md mirror) changed. Content is intentionally not carried
// (it can be ~12KB): an open profile panel refetches GET /api/agents/:id/memory.
export interface AgentMemoryChangedEvent {
  type: "agent_memory_changed";
  agentId: string;
}

// An agent's managed services (service_* tools: dev servers, watchers) changed. Like memory,
// content doesn't ride in the broadcast: an open profile panel refetches
// GET /api/agents/:id/services.
export interface AgentServicesChangedEvent {
  type: "agent_services_changed";
  agentId: string;
}

// An always-ask agent is requesting confirmation for a sensitive tool call.
export interface ToolConfirmationRequestEvent {
  type: "tool_confirmation_request";
  confirmId: string;
  channelId: string;
  agentId: string;
  agentHandle: string;
  agentName: string;
  tool: string;
  input: unknown;
}

// A pending tool confirmation was resolved (by a human, or auto-denied on timeout — no `by`).
export interface ToolConfirmationResolvedEvent {
  type: "tool_confirmation_resolved";
  confirmId: string;
  channelId: string;
  result: "allow" | "deny";
  by?: string;
}

// A schedule in the recipient's workspace changed (created/updated/deleted, including fires and
// auto-pauses, which are updates). Coarse by design: clients refetch the schedule list.
export interface ScheduleChangedEvent {
  type: "schedule_changed";
  scheduleId: string;
  action: "created" | "updated" | "deleted";
}

// A workflow in the recipient's workspace changed (created/updated/deleted — including draft
// edits by the Architect, which is what makes the builder's live preview work). Coarse by
// design, like schedule_changed: clients refetch the workflow.
export interface WorkflowChangedEvent {
  type: "workflow_changed";
  workflowId: string;
  action: "created" | "updated" | "deleted";
}

// A workflow run started or changed status (done/stalled/stopped). Clients refetch the run
// (or the workflow's run list).
export interface WorkflowRunChangedEvent {
  type: "workflow_run_changed";
  workflowId: string;
  runId: string;
}

// An agent shipped a work artifact (a PR opened, a doc written, …) — extracted from the links in
// its message at send time. Carries the full row so the Deliverables feed appends without a refetch.
export interface DeliverableCreatedEvent {
  type: "deliverable_created";
  deliverable: Deliverable;
}

// A channel's Slack mirror binding changed (linked, unlinked, or moved to the 'error' state).
// `link` is null when the channel was unlinked. Broadcast workspace-wide so every client's
// channel header updates. SlackChannelLink comes from ./slack.
export interface SlackLinkChangedEvent {
  type: "slack_link_changed";
  channelId: string;
  link: SlackChannelLink | null;
}

export type ServerEvent =
  | ConnectedEvent
  | ErrorEvent
  | PongEvent
  | MessageEvent
  | AgentStatusChangedEvent
  | DeviceStatusChangedEvent
  | MembersChangedEvent
  | ChannelDeletedEvent
  | ChannelCreatedEvent
  | ParticipantUpdatedEvent
  | ParticipantCreatedEvent
  | ParticipantDeletedEvent
  | AgentTurnEvent
  | AgentEventEvent
  | AgentQueuedEvent
  | AgentContextEvent
  | AgentMemoryChangedEvent
  | AgentServicesChangedEvent
  | ToolConfirmationRequestEvent
  | ToolConfirmationResolvedEvent
  | ScheduleChangedEvent
  | WorkflowChangedEvent
  | WorkflowRunChangedEvent
  | DeliverableCreatedEvent
  | SlackLinkChangedEvent;

// ---- Client -> server ----

// Post a message (or thread reply) to a channel. Needs a body and/or pre-uploaded
// attachmentIds. threadRootId makes it a thread reply; alsoToChannel echoes that reply into
// the main channel timeline.
export interface ClientPostFrame {
  type: "post";
  channelId: string;
  body?: string;
  clientMsgId?: string;
  attachmentIds?: string[];
  threadRootId?: string | null;
  alsoToChannel?: boolean;
}

// Application-level liveness probe; the server answers with a `pong`. Sent on a timer by the web
// client, which reconnects if the replies stop coming (see PongEvent).
export interface ClientPingFrame {
  type: "ping";
}

export type ClientFrame = ClientPostFrame | ClientPingFrame;

File: ios/Jungle/Wire/WSEvents.swift
import Foundation

// Hand-ported from shared/src/ws-events.ts — the app WebSocket contract. WS envelopes are
// camelCase on the wire (unlike the snake_case domain rows they carry). A PR touching
// ws-events.ts should touch this file.

// Where an agent's current turn came from (the dispatch the runner consumed).
struct TurnContext: Codable, Hashable, Sendable {
    var channelId: String?
    var threadRootId: String?
    var messageId: String?
}

// From shared/src/slack.ts (carried by slack_link_changed).
struct SlackChannelLink: Codable, Hashable, Sendable {
    var channelId: String
    var slackChannelId: String
    var slackChannelName: String?
    var status: String // "active" | "error"
    var lastError: String?
}

// Server -> client. One case per ServerEvent variant, plus .unknown so a new server event
// type never crashes an old app build.
enum ServerEvent: Sendable {
    case connected(participantId: String)
    case error(String)
    case message(Message)
    case agentStatusChanged(agentId: String, status: AgentStatus)
    case deviceStatusChanged(deviceId: String, online: Bool)
    case membersChanged(channelId: String)
    case channelDeleted(channelId: String)
    case participantUpdated(Participant)
    case participantDeleted(participantId: String)
    case agentTurn(agentId: String, turnId: String, context: TurnContext?)
    case agentEvent(agentId: String, turnId: String, event: JSONValue, context: TurnContext?)
    case agentQueued(agentId: String, context: TurnContext)
    case agentContext(agentId: String, tokens: Int, maxTokens: Int)
    case agentMemoryChanged(agentId: String)
    case toolConfirmationRequest(ToolConfirmation)
    case toolConfirmationResolved(confirmId: String, channelId: String, result: String, by: String?)
    case scheduleChanged(scheduleId: String, action: String)
    case deliverableCreated(Deliverable)
    case slackLinkChanged(channelId: String, link: SlackChannelLink?)
    case unknown(type: String)
}

// The tool_confirmation_request payload (also rebuilt from GET /api/confirmations).
struct ToolConfirmation: Codable, Identifiable, Hashable, Sendable {
    var confirmId: String
    var channelId: String
    var agentId: String
    var agentHandle: String
    var agentName: String
    var tool: String
    var input: JSONValue

    var id: String { confirmId }
}

extension ServerEvent: Decodable {
    private enum CodingKeys: String, CodingKey {
        case type, participantId, error, message, agentId, status, deviceId, online, channelId
        case participant, turnId, context, event, tokens, maxTokens, confirmId, result, by
        case scheduleId, action, deliverable, link
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let type = try c.decode(String.self, forKey: .type)
        switch type {
        case "connected":
            self = .connected(participantId: try c.decode(String.self, forKey: .participantId))
        case "error":
            self = .error(try c.decode(String.self, forKey: .error))
        case "message":
            self = .message(try c.decode(Message.self, forKey: .message))
        case "agent_status_changed":
            self = .agentStatusChanged(
                agentId: try c.decode(String.self, forKey: .agentId),
                status: try c.decode(AgentStatus.self, forKey: .status))
        case "device_status_changed":
            self = .deviceStatusChanged(
                deviceId: try c.decode(String.self, forKey: .deviceId),
                online: try c.decode(Bool.self, forKey: .online))
        case "members_changed":
            self = .membersChanged(channelId: try c.decode(String.self, forKey: .channelId))
        case "channel_deleted":
            self = .channelDeleted(channelId: try c.decode(String.self, forKey: .channelId))
        case "participant_updated":
            self = .participantUpdated(try c.decode(Participant.self, forKey: .participant))
        case "participant_deleted":
            self = .participantDeleted(participantId: try c.decode(String.self, forKey: .participantId))
        case "agent_turn":
            self = .agentTurn(
                agentId: try c.decode(String.self, forKey: .agentId),
                turnId: try c.decode(String.self, forKey: .turnId),
                context: try c.decodeIfPresent(TurnContext.self, forKey: .context))
        case "agent_event":
            self = .agentEvent(
                agentId: try c.decode(String.self, forKey: .agentId),
                turnId: try c.decode(String.self, forKey: .turnId),
                event: try c.decodeIfPresent(JSONValue.self, forKey: .event) ?? .null,
                context: try c.decodeIfPresent(TurnContext.self, forKey: .context))
        case "agent_queued":
            self = .agentQueued(
                agentId: try c.decode(String.self, forKey: .agentId),
                context: try c.decode(TurnContext.self, forKey: .context))
        case "agent_context":
            self = .agentContext(
                agentId: try c.decode(String.self, forKey: .agentId),
                tokens: try c.decode(Int.self, forKey: .tokens),
                maxTokens: try c.decode(Int.self, forKey: .maxTokens))
        case "agent_memory_changed":
            self = .agentMemoryChanged(agentId: try c.decode(String.self, forKey: .agentId))
        case "tool_confirmation_request":
            self = .toolConfirmationRequest(try ToolConfirmation(from: decoder))
        case "tool_confirmation_resolved":
            self = .toolConfirmationResolved(
                confirmId: try c.decode(String.self, forKey: .confirmId),
                channelId: try c.decode(String.self, forKey: .channelId),
                result: try c.decode(String.self, forKey: .result),
                by: try c.decodeIfPresent(String.self, forKey: .by))
        case "schedule_changed":
            self = .scheduleChanged(
                scheduleId: try c.decode(String.self, forKey: .scheduleId),
                action: try c.decode(String.self, forKey: .action))
        case "deliverable_created":
            self = .deliverableCreated(try c.decode(Deliverable.self, forKey: .deliverable))
        case "slack_link_changed":
            self = .slackLinkChanged(
                channelId: try c.decode(String.self, forKey: .channelId),
                link: try c.decodeIfPresent(SlackChannelLink.self, forKey: .link))
        default:
            self = .unknown(type: type)
        }
    }
}

// Client -> server: the single `post` frame (messages, thread replies, steering all go here —
// sending is WS-only, there is no REST send).
struct ClientPostFrame: Encodable, Sendable {
    var type = "post"
    var channelId: String
    var body: String?
    var clientMsgId: String?
    var attachmentIds: [String]?
    var threadRootId: String?
    var alsoToChannel: Bool?
}
